File: src/frontend/src/types/company.ts
export type CompanyStatus =
  | "applied"
  | "phone_screen"
  | "onsite"
  | "offer"
  | "rejected";

export interface Company {
  id: number;
  name: string;
  group_tag: string | null;
  interview_stages: Record<string, unknown>[];
  status: CompanyStatus;
  applied_at: string | null;
  notes: string | null;
  prep_notes: string | null;
}

export interface CompanyWithWeights extends Company {
  topic_weights: TopicWeight[];
}

export interface TopicWeight {
  node_id: number;
  node_title: string;
  weight: number;
}

export interface FocusTopic {
  node_id: number;
  title: string;
  weight: number;
  progress_pct: number;
  confidence: number;
}

export interface CompanyCreate {
  name: string;
  group_tag?: string | null;
  status?: CompanyStatus;
  applied_at?: string | null;
  notes?: string | null;
  prep_notes?: string | null;
}

export type DocKind =
  | "prep_note"
  | "hub_doc"
  | "recruiter_call"
  | "other"
  | "card_index"
  | "canonical_hub"
  | "composition"
  | "drill";

export interface CompanyDocument {
  id: number;
  company_id: number;
  title: string;
  content: string;
  source_type: string;
  doc_kind: DocKind;
  is_golden: boolean;
  golden_at: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface CardIndexProblem {
  id: number;
  leetcode_id: number | null;
  title: string;
  one_liner: string;
}

export interface CardIndexCard {
  name_zh: string;
  name_en: string;
  summary_zh: string;
  problems: CardIndexProblem[];
}

export interface CardIndexContent {
  schema_version: number;
  cards: CardIndexCard[];
}

File: src/backend/schemas/company.py
"""Pydantic schemas for Company and CompanyDocument."""
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class CompanyCreate(BaseModel):
    """Schema for creating a company."""

    name: str = Field(min_length=1)
    group_tag: str | None = None
    interview_stages: list[dict] = []
    status: Literal[
        "applied", "phone_screen", "onsite", "offer", "rejected"
    ] = "applied"
    applied_at: date | None = None
    notes: str | None = None
    prep_notes: str | None = None


class CompanyUpdate(BaseModel):
    """Schema for partial company update."""

    name: str | None = Field(default=None, min_length=1)
    group_tag: str | None = None
    interview_stages: list[dict] | None = None
    status: Literal[
        "applied", "phone_screen", "onsite", "offer", "rejected"
    ] | None = None
    applied_at: date | None = None
    notes: str | None = None
    prep_notes: str | None = None


class CompanyResponse(BaseModel):
    """Schema for company API response."""

    id: int
    name: str
    group_tag: str | None = None
    interview_stages: list[dict] = []
    status: str = "applied"
    applied_at: date | None = None
    notes: str | None = None
    prep_notes: str | None = None

    model_config = ConfigDict(from_attributes=True)


class TopicWeightCreate(BaseModel):
    """Schema for creating/updating a topic weight."""

    framework_node_id: int
    weight: float = Field(default=1.0, ge=0, le=5)


class CompanyDocumentCreate(BaseModel):
    """Schema for creating a company document."""

    title: str = Field(min_length=1)
    content: str = ""
    source_type: str = "manual"


class CompanyDocumentUpdate(BaseModel):
    """Schema for updating a company document."""

    title: str | None = None
    content: str | None = None


class CompanyDocumentResponse(BaseModel):
    """Schema for company document API response."""

    id: int
    company_id: int
    title: str
    content: str
    source_type: str
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)
